### DISCUSSION TITLE
Banks and randomly Messages

### DISCUSSION THREAD

**Discussion by WeaponX15** [tttapa/Control-Surface#595] [Q&A]
Hi Peter, this is the last thing i have to do for finish; i use ManyButtonsSelector class to select especific bank, but i need the midi messages be specific c# by channel/bank:

I have 4 bankable ccbuttons to 4 banks, and those buttons are linked/cloned with others "ccbuttons line code" to send cc# messages to assign others functions, when i push this buttons sends ramdonmly the follow messages depending the active bank:

Push Button 1: sends messages CH1-C1, or CH2-C1, or CH3-C1, or CH4-C1
Push Button 2: sends messages CH1-C2, or CH2-C2, or CH3-C2, or CH4-C2
Push Button 3: sends messages CH1-C3, or CH2-C3, or CH3-C3, or CH4-C3
Push Button 4: sends messages CH1-C4, or CH2-C4, or CH3-C4, or CH4-C4

The problem is that this c# mesagges its not always the same, for example standing on Bank 2 (Channel 2), i push button 1 and can sends ramdomly CH2-C1 or CH1-C1 and i need will be same message always standing in bank 2 (CH2-C2), and once the button is released after that send messages from bank 1. I need only when changing channels to keep the channel from where it comes.

There is some way of keep only active bank just at moment pulse the button? For example sends this especific messages staying in Channel 4 (bank 4) and changes bank too:

Standing in bank 4:
Push Button 1: send message CH4-C1 (jump to bank 1)
Push Button 2: send message CH4-C2 (jump to bank 2)
Push Button 3: send message CH4-C3 (jump to bank 3)
Push Button 4: send message CH4-C4 (keep bank 4)

Standing in bank 3:
Push Button 1: send message CH3-C1 (jump to bank 1)
Push Button 2: send message CH3-C2 (jump to bank 2)
Push Button 3: send message CH3-C3 (keep bank 3)
Push Button 4: send message CH3-C4 (jump to bank 4)

etc...

I need to sending specific C#-CH# messages because i assigned diferent functions in differents channels/banks to those banks buttons.

Thanks Peter again.

**Comment by tttapa** [comment]
It's not entirely clear to me what you want to achieve, but I probably wouldn't try to fit it into the existing code. It'll be much easier and flexible to read the buttons yourself (e.g. using the included `Button` class) and then send the MIDI messages yourself using `Control_Surface.sendControlChange(address, value)`.

**Reply by WeaponX15** [reply]
What I want to do is that when I pressed the bank change button, another message was also sent at the same time to be mapped and activate another function that is mapped, the problem is that the message that is sent is random, never is the same, and I need to always respect the channel where I am, example:

What I want to do:
Being on channel 1:
I press button 1: message sent = CH1-C1

What is currently happening and I want to fix:
I press button 1: message sent = CH1-C1 or CH2-C1 (I need this message to always be the same and not change)

**Reply by tttapa** [reply]
This is to be expected, the order in which the button press is registered (i.e. by the CCButton or by the ManyButtonsSelector) is nondeterministic. 

You cannot use a ManyButtonsSelector and a CCButton on the same pin if the CCButton address depends on the selection of the ManyButtonsSelector.  
You'll have to write a class or a function that only reads the pin once, sends the CC message, and then changes the bank, in a deterministic order.
